Fix for closure-captured values not being dropped on panic

ArturKovacs opened the pull request:
- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

Fixes #1850

This moves the ownership of the callback-closure to the `EventLoop`. The colusre is now contained within an `Rc` and the global handler only has a `Weak` reference to it, in order to ensure that the closure gets dropped with the `EventLoop`. Furthermore this catches panic unwinding, coming from the user callback to prevent it from unwinding into foreign code. The application is then requested to stop and the caught unwinding is resumed at a point where the callstack doesn't contain foreign code.

ArturKovacs commented:
@francesca64 do you mind giving this a look? Also I don't want to necessarily always ask you to be the "other maintainer" when reviewing macOS PRs but I don't know of any other maintainer who has experience with macOS.

maroider commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@
 - On Android, unimplemented events are marked as unhandled on the native event loop.
 - On Windows, added `WindowBuilderExtWindows::with_menu` to set a custom menu at window creation time.
 - On Android, bump `ndk` and `ndk-glue` to 0.3: use predefined constants for event `ident`.
+- On macOS, fix event-closure-captured objects not being dropped on panic.
```
```suggestion
- On macOS, fix objects captured by the event loop closure not being dropped on panic.
```
I'm not too sure if this reads better to everyone, but it does to me.

ArturKovacs replied:
Yes, this seems easier to understand.

maroider reviewed (changes requested):
Hello!
I come bearing nit-picks!

---

On a more serious note: one thing that does worry me here is whether it is possible to panic across FFI boundaries from the callback.

ArturKovacs commented on src/platform_impl/macos/app_state.rs:
```diff
@@ -68,23 +68,43 @@ impl<T> Debug for EventLoopHandler<T> {
+                *control_flow = ControlFlow::Exit;
+            }
+        } else {
+            // Logging an error instead of panicing because might happen while the
+            // application is already panicing.
```
Thanks!

ArturKovacs commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -50,6 +55,16 @@ impl<T: 'static> EventLoopWindowTarget<T> {
+    /// into a strong reference in order to call the callback but then the
+    /// strong reference should be dropped as soon as possible.
+    callback:
+        Option<Arc<Mutex<Box<dyn FnMut(Event<'_, T>, &RootWindowTarget<T>, &mut ControlFlow)>>>>,
```
My assumption was that this has to be `Arc<Mutex>` because the callback is exposed through the global `HANDLER` object. But apparently the borrow checker is still happy if this is just `Rc<RefCell>` which is a complete mystery to me and it makes be a bit worried.

francesca64 replied:
I _vaguely_ remember there being something particular about `HANDLER` only being accessed on one thread, but this was over 2 years ago and I frankly don't have a mental model of this anymore.

I optimistically had a reason for putting non-thready things in `EventLoopHandler`, and apparently I also reviewed a PR that furthered that decision: #1108 

I really don't know anymore, though.

ArturKovacs commented:
> I come bearing nit-picks!

Welcome here :D

> On a more serious note: one thing that does worry me here is whether it is possible to panic across FFI boundaries from the callback.

Unfortunately it is. The stack trace for events on macOS is approximately
```
- "event handler callback"
- app_state::EventLoopHandler::handle_nonuser_event
- app_state::Handler::handle_nonuser_event
- app_state::AppState::cleared
- observer::control_flow_end_handler
- "foreign code"
...
- event_loop::EventLoop::run
```

Although when I tested, it seemed to work as expected but considering that it's undefined it's not exactly a solution. As you noted earlier, this ties back into #141. I'll give this a bit more thought.

maroider commented:
The Windows backend's `EventLoopRunner` may serve as a source of inspiration.

maroider commented:
Is there anything stopping you from storing it in the `EventLoop`?

francesca64 commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -114,6 +164,49 @@ impl<T> EventLoop<T> {
+                // Posting an dummy event to get stop to take effect immediately.
+                // See: https://stackoverflow.com/questions/48041279/stopping-the-nsapplication-main-event-loop/48064752#48064752
+                let event_class = class!(NSEvent);
+                let dummy_event: id = msg_send![
```
It looks like there's code duplication here with the dummy event in `AppState::cleared`, which would be great to consolidate!

francesca64 commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -114,6 +164,49 @@ impl<T> EventLoop<T> {
+            unsafe {
+                let app_class = class!(NSApplication);
+                let app: id = msg_send![app_class, sharedApplication];
+                let _: () = msg_send![app, stop: nil];
```
I know this isn't consistently used yet, but this is equivalent:
```suggestion
                let () = msg_send![app, stop: nil];
```

francesca64 commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -50,6 +65,16 @@ impl<T: 'static> EventLoopWindowTarget<T> {
+    /// Every other reference should be a Weak reference which is only upgraded
+    /// into a strong reference in order to call the callback but then the
+    /// strong reference should be dropped as soon as possible.
+    _callback:
```
What's the motive for prefixing this with an underscore? It looks like this field is accessed below.

ArturKovacs replied:
The nightly compiler emits a warning, because this is never read. https://github.com/rust-windowing/winit/pull/1853/checks?check_run_id=1892520990#step:9:63

francesca64 replied:
Ah, gotcha! Thanks for improving my understanding of the rules around that.

francesca64 reviewed (changes requested):
Looks great! I won't be able to test it out until next week, though.

Thanks so much for fixing this!

> I don't want to necessarily always ask you to be the "other maintainer" when reviewing macOS PRs but I don't know of any other maintainer who has experience with macOS.

No worries! You can always ping me, but know that I don't have a ton of bandwidth for winit, so you'll get to hear me apologize for delays a bunch!

I'm not aware of anyone else around who's familiar with macOS, though @kchibisov has reviewed macOS PRs in the past. In any case, feel free to let me know if you see anyone around who would be good to make a Member too!

ArturKovacs commented:
> Is there anything stopping you from storing it in the `EventLoop`?

I moved it into the `EventLoop`. To ensure that the conversion from the raw pointer (at `observer::control_flow_end_handler`) into a rust object is safe, I used `Rc` and `Weak` here as well. Earlier we had a "Mutex locking" which we replaced with upgrading a `Weak` to an `Rc`. Considering that the Mutex (in theory) never experienced contention, I believe that this improvement is negligible regarding runtime speed. But now it's done so there you go.

Although there's one thing I'm entirely sure about. I had to use `AssertUnwindSafe` on the `Weak` because the standard library doesn't have a blanket implementation for `UnwindSafe` for `Weak<T>` even if T implements `RefUnwindSafe`. Whereas `Rc` does have this blanket implementation so I'm a bit confused.

> I'm not aware of anyone else around who's familiar with macOS, though @kchibisov has reviewed macOS PRs in the past.

Thanks!

> In any case, feel free to let me know if you see anyone around who would be good to make a Member too!

Will do!

ArturKovacs commented:
@francesca64 Let me know approximately when you are available, and then I'll fix the conflicts so that you can review this.

francesca64 commented on src/platform_impl/macos/observer.rs:
```diff
@@ -152,31 +192,41 @@ impl RunLoop {
 }
 
-pub fn setup_control_flow_observers() {
+pub(crate) fn setup_control_flow_observers(panic_info: Weak<PanicInfo>) {
```
This function isn't visible outside of `platform_impl::macos` either way, right? I'm not sure if adding `(crate)` makes that more explicit or if it makes the overall module structure and their relative visibilities less clear.

(This comment applies to other things this PR marks `pub(crate)` too)

ArturKovacs replied:
As far as I understand, if these are just `pub` then they can still be reached from outside the crate. This also limits the optimizations that can be done when compiling the crate.

But I don't feel strongly about this, so I'm changing them to `pub`.

francesca64 commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -114,6 +191,54 @@ impl<T> EventLoop<T> {
+                let app: id = msg_send![app_class, sharedApplication];
+                let () = msg_send![app, stop: nil];
+
+                // Posting an dummy event to get stop to take effect immediately.
```
Using backticks makes it easier not to read "stop" as a verb:
```suggestion
                // Posting a dummy event to get `stop` to take effect immediately.
```

francesca64 commented on src/platform_impl/macos/app_state.rs:
```diff
@@ -265,8 +281,9 @@ impl AppState {
-    pub fn wakeup() {
-        if !HANDLER.is_ready() {
+    pub fn wakeup(panic_info: Weak<PanicInfo>) {
+        let panic_info = panic_info.upgrade().unwrap();
```
If these `unwrap`s are more-or-less unreachable, then making them `expect`s that indicate that this panicking is a developer error / invariant violation would be great.

francesca64 commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -23,6 +33,38 @@ use crate::{
+}
+// ------------------------------------
+// WARNING:
+// This struct is UnwindSafe if `get_mut` is never called on the inner data.
```
If I'm reading this right, we're safe from this as long as no `get_mut` usages are in that `PanicInfo` impl block below, or is the scope larger than that?

ArturKovacs replied:
I think you are reading this right. Of course, the rest of the module can also directly access the private `inner` so that's why I choose the specific wording. Although I can see that it's a strange way to say what I'm trying to say here, so I just made an attempt at improving this.

francesca64 replied:
This reads very clearly now, thanks!

francesca64 commented on src/platform_impl/macos/event_loop.rs:
```diff
@@ -98,14 +152,37 @@ impl<T> EventLoop<T> {
+            >(Rc::new(RefCell::new(callback)))
+        };
+
+        self._callback = Some(callback.clone());
```
Consistency:
```suggestion
        self._callback = Some(Rc::clone(&callback));
```

francesca64 reviewed (changes requested):
I've got some deadlines looming, so I won't be around much this month. I'm taking today to catch up on my github backlog though, so sorry for reviewing this before giving you a chance to fix conflicts/etc.

ArturKovacs commented on src/platform_impl/macos/observer.rs:
```diff
@@ -103,41 +124,60 @@ pub struct CFRunLoopSourceContext {
     pub perform: Option<extern "C" fn(*mut c_void)>,
 }
 
+fn control_flow_handler<F>(panic_info: *mut c_void, f: F)
```
It damn well does :D

ArturKovacs commented on src/platform_impl/macos/observer.rs:
```diff
@@ -1,6 +1,18 @@
-use std::{self, os::raw::*, ptr, time::Instant};
+use core::panic;
```
It undeniably seems that way :D

ArturKovacs commented on src/platform_impl/macos/observer.rs:
```diff
@@ -152,31 +192,41 @@ impl RunLoop {
     unsafe {
+        let mut context = CFRunLoopObserverContext {
+            info: Weak::into_raw(panic_info) as *mut _,
+            version: 0,
```
Yeah putting the documentation there makes sense.

francesca64 reviewed (approved):
Everything looks awesome! Thanks again for fixing this.

(I also saw the reddit thread you made about `Weak`'s unwind safety; you have great initiative!)

ArturKovacs commented:
> (I also saw the reddit thread you made about `Weak`'s unwind safety; you have great initiative!)

Oh stop it, you 😊
